controllers/dspipeline_controller.go

```diff
@@ -189,46 +222,116 @@ func (r *DSPAReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
 	err = r.ReconcileStorage(ctx, dspa, params)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	if r.isDeploymentAvailable(ctx, dspa, "minio") {
+		objectStorageReady.Status = metav1.ConditionTrue
+	}
+	conditions = append(conditions, objectStorageReady)
+
 	err = r.ReconcileCommon(dspa, params)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	apiServerReady := r.buildCondition(config.APIServerReady, dspa, config.MinimumReplicasAvailable)
 	err = r.ReconcileAPIServer(ctx, dspa, params)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	if r.isDeploymentAvailable(ctx, dspa, "ds-pipeline") {
+		apiServerReady.Status = metav1.ConditionTrue
+	}
+	conditions = append(conditions, apiServerReady)
+
+	persistenceAgentReady := r.buildCondition(config.PersistenceAgentReady, dspa, config.MinimumReplicasAvailable)
 	err = r.ReconcilePersistenceAgent(dspa, params)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	if r.isDeploymentAvailable(ctx, dspa, "ds-pipeline-persistenceagent") {
+		persistenceAgentReady.Status = metav1.ConditionTrue
+	}
+	conditions = append(conditions, persistenceAgentReady)
+
+	scheduledWorkflowReady := r.buildCondition(config.ScheduledWorkflowReady, dspa, config.MinimumReplicasAvailable)
 	err = r.ReconcileScheduledWorkflow(dspa, params)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	if r.isDeploymentAvailable(ctx, dspa, "ds-pipeline-scheduledworkflow") {
+		scheduledWorkflowReady.Status = metav1.ConditionTrue
+	}
+	conditions = append(conditions, scheduledWorkflowReady)
+
+	userInterfaceReady := r.buildCondition(config.UserInterfaceReady, dspa, config.MinimumReplicasAvailable)
 	err = r.ReconcileUI(dspa, params)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	if r.isDeploymentAvailable(ctx, dspa, "ds-pipeline-ui") {
+		userInterfaceReady.Status = metav1.ConditionTrue
+	}
+	conditions = append(conditions, userInterfaceReady)
+
 	err = r.ReconcileViewerCRD(dspa, params)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	log.Info("Updating CR status")
+
+	crReady := r.buildCondition(config.CrReady, dspa, config.MinimumReplicasAvailable)
+	crReady.Type = config.CrReady
+
+	// Compute Ready Logic for the CR
+	if ((apiServerReady.Status == metav1.ConditionTrue) &&
+		(persistenceAgentReady.Status == metav1.ConditionTrue) &&
+		(scheduledWorkflowReady.Status == metav1.ConditionTrue) &&
+		(databaseReady.Status == metav1.ConditionTrue) &&
+		(objectStorageReady.Status == metav1.ConditionTrue)) &&
+		(userInterfaceReady.Status == metav1.ConditionTrue || (userInterfaceReady.Status == metav1.ConditionFalse && !dspa.Spec.MlPipelineUI.Deploy)) {
+		crReady.Status = metav1.ConditionTrue
+	} else {
+		crReady.Status = metav1.ConditionFalse
+	}
+	conditions = append(conditions, crReady)
+
+	for i, condition := range dspa.Status.Conditions {
+		if condition.Status != conditions[i].Status {
+			conditions[i].LastTransitionTime = metav1.Now()
+		}
```

Comment:
You need to handle the case where the status has not changed. 

To avoid this you can do the following condition check:

```suggestion
		if condition.Status != conditions[i].Status {
			conditions[i].LastTransitionTime = metav1.Now()
		} else {
			conditions[i].LastTransitionTime = condition.LastTransitionTime
		}
```

Without this, you will get the following error: 

```
2023-04-11T16:09:32-04:00       INFO    DataSciencePipelinesApplication.datasciencepipelinesapplications.opendatahub.io "sample" is invalid: [status.conditions[0].lastTransitionTime: Required value, status.conditions[1].lastTransitionTime: Required value, status.conditions[2].lastTransitionTime: Required value, status.conditions[3].lastTransitionTime: Required value, status.conditions[4].lastTransitionTime: Required value, status.conditions[5].lastTransitionTime: Required value, status.conditions[6].lastTransitionTime: Required value]        {"namespace": "test-project"}
2023-04-11T16:09:32-04:00       ERROR   Reconciler error        {"controller": "datasciencepipelinesapplication", "controllerGroup": "datasciencepipelinesapplications.opendatahub.io", "controllerKind": "DataSciencePipelinesApplication", "DataSciencePipelinesApplication": {"name":"sample","namespace":"test-project"}, "namespace": "test-project", "name": "sample", "reconcileID": "d00e5863-5c18-4e52-a349-6c7212bba316", "error": "DataSciencePipelinesApplication.datasciencepipelinesapplications.opendatahub.io \"sample\" is invalid: [status.conditions[0].lastTransitionTime: Required value, status.conditions[1].lastTransitionTime: Required value, status.conditions[2].lastTransitionTime: Required value, status.conditions[3].lastTransitionTime: Required value, status.conditions[4].lastTransitionTime: Required value, status.conditions[5].lastTransitionTime: Required value, status.conditions[6].lastTransitionTime: Required value]"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /home/hukhan/go/1.19.5/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /home/hukhan/go/1.19.5/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /home/hukhan/go/1.19.5/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:234

```